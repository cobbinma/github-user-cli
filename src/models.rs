use serde::export::Formatter;

#[derive(Default, Debug, Clone, PartialEq, serde_derive::Serialize, serde_derive::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Repository {
    pub name: String,
    pub description: Option<String>,
    pub url: String,
    #[serde(rename = "stargazers_count")]
    pub stars: i64,
}

impl std::fmt::Display for Repository {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "name: '{}', ", self.name)?;
        if let Some(d) = &self.description {
            write!(f, "description: '{}', ", d)?;
        };
        write!(f, "url: '{}', ", self.url)?;
        write!(f, "stars: '{}'", self.stars)
    }
}
