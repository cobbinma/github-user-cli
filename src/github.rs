use crate::models::{Repository, RepositoryClient};
use async_trait::async_trait;
use std::error::Error;

pub struct GitHub {
    token: Option<String>,
}

impl GitHub {
    pub fn new(token: Option<String>) -> Box<GitHub> {
        Box::new(GitHub { token })
    }
}

#[async_trait]
impl RepositoryClient for GitHub {
    async fn get_repositories(&self, username: &str) -> Result<Vec<Repository>, Box<dyn Error>> {
        surf::get(format!("https://api.github.com/users/{}/repos", username))
            .recv_json()
            .await
            .map_err(From::from)
    }
}
