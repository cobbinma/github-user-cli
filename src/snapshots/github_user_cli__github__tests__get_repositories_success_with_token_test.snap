---
source: src/github.rs
expression: "&repositories"
---
[
  {
    "name": "r-cache",
    "description": "r-cache is an in memory key value store. It is thread safe and values can have expiry times",
    "html_url": "https://github.com/cobbinma/r-cache",
    "stargazers_count": 6
  },
  {
    "name": "example-go-api",
    "description": "Example Go REST API",
    "html_url": "https://github.com/cobbinma/example-go-api",
    "stargazers_count": 63
  }
]
